### DISCUSSION TITLE
vcpkg, cmake and bullet3

### DISCUSSION THREAD

**Discussion by AshSinc** [microsoft/vcpkg#19172] [Q&A] [answered]
Hi,

I'm trying to use vcpkg's bullet package, and I'm beginning to wonder if some of my problems are stemming from the vcpkg bullet package itself.

I've made a StackOverflow post [https://stackoverflow.com/questions/68547187/cmake-vcpkg-and-bullet](url) perhaps with more information.

Basically I've followed the installation procedure on the bullet3 git repo:
```
git clone https://github.com/Microsoft/vcpkg.git
cd vcpkg
./bootstrap-vcpkg.sh
./vcpkg integrate install
./vcpkg install bullet3
```
Then I find the usage instructions for CMake:
```
The package bullet3 provides CMake targets:
find_package(Bullet CONFIG REQUIRED)
target_link_libraries(main PRIVATE LinearMath Bullet3Common BulletDynamics BulletSoftBody)
```

After adding them to my CMakeLists.txt I get an error:

> Unknown CMake command "set_and_check"

In BulletConfig.cmake I find these set_and_check calls and the linter doesnt recognise them. I find references to the function online and in the CMake docs. But despite having the latest version of CMake I don't have access to this command. Was it deprecated at some point?

I change them to set() instead and with a little debugging find that PACKAGE_PREFIX_DIR variable is not populated. So i statically set the PACKAGE_PREFIX_DIR variable to package path. This leaves me with this BulletConfig.cmake file.

```
#                                               -*- cmake -*-
#
#  BulletConfig.cmake(.in)
#

# Use the following variables to compile and link against Bullet:
#  BULLET_FOUND              - True if Bullet was found on your system
#  BULLET_USE_FILE           - The file making Bullet usable
#  BULLET_DEFINITIONS        - Definitions needed to build with Bullet
#  BULLET_INCLUDE_DIR        - Directory where Bullet-C-Api.h can be found
#  BULLET_INCLUDE_DIRS       - List of directories of Bullet and it's dependencies
#  BULLET_LIBRARIES          - List of libraries to link against Bullet library
#  BULLET_LIBRARY_DIRS       - List of directories containing Bullet' libraries
#  BULLET_ROOT_DIR           - The base directory of Bullet
#  BULLET_VERSION_STRING     - A human-readable string containing the version

set(PACKAGE_PREFIX_DIR /home/ash/installed/x64-linux)

set ( BULLET_FOUND 1 )
set_and_check ( BULLET_USE_FILE     "${PACKAGE_PREFIX_DIR}/share/bullet3/UseBullet.cmake" )
set ( BULLET_DEFINITIONS  "" )
set ( BULLET_INCLUDE_DIR  "${PACKAGE_PREFIX_DIR}/include/bullet" )
set ( BULLET_INCLUDE_DIRS "${PACKAGE_PREFIX_DIR}/include/bullet" )
set ( BULLET_LIBRARIES    "LinearMath;Bullet3Common;BulletInverseDynamics;BulletCollision;BulletDynamics;BulletSoftBody" )
set ( BULLET_LIBRARY_DIRS "${PACKAGE_PREFIX_DIR}/lib" )
set ( BULLET_ROOT_DIR     "${PACKAGE_PREFIX_DIR}" )
set ( BULLET_VERSION_STRING "3.17" )

# Load targets
if(NOT TARGET Bullet3Common)
  file(GLOB CONFIG_FILES "${PACKAGE_PREFIX_DIR}/share/bullet3/*Targets.cmake")
  foreach(f ${CONFIG_FILES})
    include(${f})
  endforeach()
  set(_DIR)
endif()
```

With this the variables BULLET_INCLUDE_DIR etc are populated with the correct paths. CMake builds the makefile. But running make i get :

> fatal error: btBulletDynamicsCommon.h: No such file or directory
> #include <btBulletDynamicsCommon.h>

So I could well be (i am) doing something wrong. But I'm in need of help. So first of all is this the intended behaviour of the vcpkg bullet package? And have I gone in the wrong direction somewhere?

Or is there an issue with the package just now?

Any help will be so appreciated. Thank you.

EDIT1

I created an empty project and ran through each library I wanted to include. Everything works except for Bullet3. However it does now see the header files. What changed between the two CMakeFiles? Nothing as far as I can tell. I'll need to find out because I have to port this project over but in the meantime this is another issue with the package.

>     from /home/ash/projects/C++/CMakeImportTests/src/main.cpp:22:
>     /home/ash/vcpkg/installed/x64-linux/include/bullet/BulletCollision/CollisionDispatch/btCollisionWorld.h:77:10: fatal error: LinearMath/btVector3.h: No such file or directory
>     77 | #include "LinearMath/btVector3.h"

I think this is the same issue as described https://github.com/microsoft/vcpkg/issues/7877

If i remove all includes of Bullet but leave the CMakeList.txt untouched, we get this error:

>     [ 50%] Building CXX object CMakeFiles/main.dir/src/main.cpp.o
>     [100%] Linking CXX executable main
>     /usr/bin/ld: cannot find -lLinearMath
>     /usr/bin/ld: cannot find -lBullet3Common
>     /usr/bin/ld: cannot find -lBulletDynamics
>     /usr/bin/ld: cannot find -lBulletSoftBody
>     collect2: error: ld returned 1 exit status
>     make[2]: *** [CMakeFiles/main.dir/build.make:104: main] Error 1
>     make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/main.dir/all] Error 2
>     make: *** [Makefile:91: all] Error 2

Is this a clue that some environment variable is not set?

EDIT2

There seems to be an ordering dependency for the target_link_library call. The suggested usage is:

`target_link_libraries(main PRIVATE LinearMath Bullet3Common BulletDynamics BulletSoftBody)`

Checking bullet.pc in the libs/ directory i found 

`Libs: -L${libdir} -lBulletSoftBody -lBulletDynamics -lBulletCollision -lLinearMath`

So I tried rearranging and following the pattern:

`target_link_libraries(main PRIVATE BulletSoftBody BulletDynamics BulletCollision Bullet3Common LinearMath)`

Additionally there was also a need to manually link directories.

`target_link_directories(main PRIVATE ${BULLET_LIBRARY_DIRS})`

This now compiles without error in my test project. It seems LinearMath must be after most of the other libraries (although it can be before Bullet3Common it seems).

For some reason it's still not finding the header files when I copy the exact same CMake commands over to my main project. So I'm not free of this yet.

I should say that I was able to remove the change I made to BulletConfig.cmake of setting PACKAGE_PREFIX_DIR statically.

So just to recap my issue. A small test project works and I can use bullet and number of other libraries that I use in my main project. But if i copy this working CMakeLists.txt to my main project it can no longer find the headers and throws this error :

> btBulletDynamicsCommon.h: No such file or directory
>     8 | #include <btBulletDynamicsCommon.h>

Bullet_DIR:PATH=/home/ash/vcpkg/installed/x64-linux/share/bullet3 is the same in both cases.

**Comment by mathisloge** [comment]
(this isn't a fix for your problem, just mentioning that )
this has been previously answered on discord. 

Bullet needs to be updated in vcpkg. 

A quick fix was mentioned by @vicroms to replace  `set_and_check` with `set`.

**Comment by mathisloge** [comment]
could you check https://github.com/microsoft/vcpkg/pull/19236 ? Should fix your problems

The important thing to do is to add `include(${BULLET_USE_FILE})` after the `find_package(Bullet ...)`

**Reply by AshSinc** [reply]
I ended up using target_include_directories(main PRIVATE ${BULLET_INCLUDE_DIR}) and removing set_and_check() calls. And changing the ordering of the target_link_libraries(). Then after a restart it just started working...

It was such a headache to fix! And in the end i had done everything correctly there was just something lingering causing a problem I guess.

I can try 19236 if it would still be useful. I'm pretty busy just now and I'm weary of breaking things again :)

Looking at the commit it seems set_and_check() is still there. What is that method?

**Reply by mathisloge** [reply]
set_and_check will be added by `@PACKAGE_INIT@` and will be replaced by cmake with the following 
```
macro(set_and_check _var _file)
  set(${_var} "${_file}")
  if(NOT EXISTS "${_file}")
    message(FATAL_ERROR "File or directory ${_file} referenced by variable ${_var} does not exist !")
  endif()
endmacro()
```
You can get more info about it https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html 

I think 19236 is currently good. If it is merged, you can revert your changes and pull the updates. 
It also fixes some issues with debug and release mode.

**Comment by JackBoosY** [comment, marked as answer]
https://github.com/microsoft/vcpkg/pull/19236 solved this.
